QEMU: Add support to qpc script for multiple os

Haroon-Khel opened the pull request:
ref https://github.com/AdoptOpenJDK/openjdk-infrastructure/issues/1538

I will be creating more qemu disk images, which will involve different OSs for already supported architectures.

Im currently adding a ubuntu 18 on arm64 qemu disk image onto `infra-softlayer-ubuntu1804-x64-1`. To accommodate the changes I have made to the script, I will rename the xz compressed disk images to support the `OS.ARCHITECTURE.dsk.xz` style.

Haroon-Khel commented:
@Willsparker Do you know the OS for the riscv disk image? All I get is Debian from 
```
10:37:04     "msg": [
10:37:04         "inventory_hostname: localhost ", 
10:37:04         "ansible_hostname: debian", 
10:37:04         "ansible_fqdn: debian", 
10:37:04         "ansible_user: linux", 
10:37:04         "ansible_default_ipv4.address: 10.0.2.15", 
10:37:04         "ansible_os_family: Debian ", 
10:37:04         "ansible_distribution: Debian ", 
10:37:04         "ansible_distribution_major_version: unstable ", 
10:37:04         "ansible_architecture: riscv64 ", 
10:37:04         "ansible_processor_vcpus: 1 ", 
10:37:04         "ansible_processor_cores: 1 ", 
10:37:04         "Domain: adoptopenjdk.net ", 
10:37:04         "Jenkins_Username: jenkins ", 
10:37:04         "Superuser_Account: Enabled", 
10:37:04         "Vendor_File: Disabled", 
10:37:04         "Nagios_Plugins: Enabled ", 
10:37:04         "Nagios_Monitoring: Enabled", 
10:37:04         "Nagios_Master_IP: 78.47.239.96"
10:37:04     ]
```
but no version number

Haroon-Khel commented:
Nevermind, from the qemu docs I see that its Debian Buster

Willsparker commented:
@Haroon-Khel I believe it's Debian 'bullseye', which will be the one after buster, AFAIK

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -94,6 +98,18 @@ defaultVars() {
+		"debian10" | "Debian10" | "deb10" )
+			echo "debian10 selected for $ARCHITECURE"; OS=DEBIAN10;;
+		"ubuntu18" | "u18" | "Ubuntu18" )
+			echo "ubuntu18 selected for $ARCHITECURE"; OS=UBUNTU18;;
```
Sorry to be nitpick-y but when `echo`ing to the user, could you capitalise the distribution?

sxa replied:
```suggestion
			echo "ubuntu18 selected for $ARCHITECTURE"; OS=UBUNTU18;;
```

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -72,6 +75,7 @@ usage() {
 		--infra-branch | -ib		Specify the branch of the infra-repo (default: master)
 		--jdk-version | -v		Specify which JDK to build if '-b' is used (default: jdk8u)
 		--retainVM | -r			Retain the VM once running the playbook
+		--operating-system | -o Combined with --architecture runs a VM with the desired architecture and OS combo.
```
```suggestion
		--operating-system | -o		Combined with --architecture runs a VM with the desired architecture and OS combo.
```

Willsparker reviewed (changes requested):
This also allows for incorrect OSs with the given architectures (i.e. Debian10/s390x is valid). 
Off the top of my head, a way to go about this would be to validate the OS/architecture combination by seeing if `$OS.ARCHITECTURE.dsk` is available in `$imageLocation`.

sxa commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -94,6 +98,18 @@ defaultVars() {
+		"debian8" | "Debian8" | "deb8" )
+			echo "debian8 selected for $ARCHITECTURE"; OS=DEBIAN8;;
+		"debian10" | "Debian10" | "deb10" )
+			echo "debian10 selected for $ARCHITECURE"; OS=DEBIAN10;;
```
```suggestion
			echo "debian10 selected for $ARCHITECTURE"; OS=DEBIAN10;;
```

sxa commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -160,30 +181,38 @@ done
 		"S390X" )
 			export MACHINE="s390-ccw-virtio"
-			export DRIVE="-drive file=$workFolder/${ARCHITECTURE}.dsk,if=none,id=hd0 -device virtio-blk-ccw,drive=hd0,id=virtio-disk0"
+			export DRIVE="-drive file=$workFolder/$OS.${ARCHITECTURE}.dsk,if=none,id=hd0 -device virtio-blk-ccw,drive=hd0,id=virtio-disk0"
```
For syntax consistency
```suggestion
			export DRIVE="-drive file=$workFolder/${OS}.${ARCHITECTURE}.dsk,if=none,id=hd0 -device virtio-blk-ccw,drive=hd0,id=virtio-disk0"
```

sxa commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -160,30 +181,38 @@ done
 		"PPC64LE" )
 			export MACHINE="pseries-2.12"
-			export DRIVE="-hda $workFolder/${ARCHITECTURE}.dsk"
+			export DRIVE="-hda $workFolder/$OS.${ARCHITECTURE}.dsk"
```
For syntax consistency
```suggestion
			export DRIVE="-hda $workFolder/${OS}.${ARCHITECTURE}.dsk"
```

sxa commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -160,30 +181,38 @@ done
 			export QEMUARCH="riscv64"
 			export MACHINE="virt"
-			export DRIVE="-device virtio-blk-device,drive=hd -drive file=$workFolder/${ARCHITECTURE}.dsk,if=none,id=hd"
+			export DRIVE="-device virtio-blk-device,drive=hd -drive file=$workFolder/$OS.${ARCHITECTURE}.dsk,if=none,id=hd"
```
For syntax consistency
```suggestion
			export DRIVE="-device virtio-blk-device,drive=hd -drive file=$workFolder/${OS}.${ARCHITECTURE}.dsk,if=none,id=hd"
```

sxa commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -160,30 +181,38 @@ done
 			export QEMUARCH="arm"
 			export SSH_CMD="-device virtio-net-device,netdev=mynet -netdev user,id=mynet,hostfwd=tcp::$PORTNO-:22"
-			export DRIVE="-drive if=none,file=$workFolder/ARM32.dsk,format=qcow2,id=hd -device virtio-blk-device,drive=hd"
+			export DRIVE="-drive if=none,file=$workFolder/$OS.$ARCHITECTURE.dsk,format=qcow2,id=hd -device virtio-blk-device,drive=hd"
```
For syntax consistency
```suggestion
			export DRIVE="-drive if=none,file=$workFolder/${OS}.${ARCHITECTURE}.dsk,format=qcow2,id=hd -device virtio-blk-device,drive=hd"
```

Haroon-Khel commented:
@Willsparker Ive put the RISCV image's distro down as debian 11, according to https://wiki.debian.org/DebianBullseye, do you object?

Haroon-Khel commented:
At https://ci.adoptopenjdk.net/job/QEMUPlaybookCheck/ Im running a QPC job with modified parameters to allow the inclusion of an OS parameter (it looks quite ugly so I will be looking for alternatives)

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -81,19 +85,33 @@ usage() {
+		"debian11" | "deb11" | "Debian11" )
+			echo "DEBIAN11 selected for $ARCHITECTURE"; OS=DEBIAN11;;
+		* )
+			echo "Please use the -o flag to select a supported OS"; showArchList; exit 1;;
```
As the `-o` flag is required, could you update the `usage` function, just so it says:
```
	echo "Usage: ./qemu_test_script.sh (<options>) -a <architecture> -o <os>
```
I would've given that as a  'suggested change', but GitHub won't let you do that for pieces of code that haven't been affected  -_-

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -81,19 +85,33 @@ usage() {
+
+	case "$OS" in
+		"debian8" | "Debian8" | "deb8" )
+			echo "DEBIAN8 selected for $ARCHITECTURE"; OS=DEBIAN8;;
```
```suggestion
			echo "Debian8 selected for \$ARCHITECTURE"; OS=DEBIAN8;;
```

Haroon-Khel replied:
Ive kept the os in caps to keep it consistent with the how the architecture is displayed

Willsparker replied:
Ah, right! Okay, yep that makes more sense

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -81,19 +85,33 @@ usage() {
+		"debian8" | "Debian8" | "deb8" )
+			echo "DEBIAN8 selected for $ARCHITECTURE"; OS=DEBIAN8;;
+		"debian10" | "Debian10" | "deb10" )
+			echo "DEBIAN10 selected for $ARCHITECTURE"; OS=DEBIAN10;;
```
```suggestion
			echo "Debian10 selected for \$ARCHITECTURE"; OS=DEBIAN10;;
```

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -81,19 +85,33 @@ usage() {
+		"debian10" | "Debian10" | "deb10" )
+			echo "DEBIAN10 selected for $ARCHITECTURE"; OS=DEBIAN10;;
+		"ubuntu18" | "u18" | "Ubuntu18" )
+			echo "UBUNTU18 selected for $ARCHITECTURE"; OS=UBUNTU18;;
```
```suggestion
			echo "Ubuntu18 selected for \$ARCHITECTURE"; OS=UBUNTU18;;
```

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -115,12 +133,18 @@ defaultVars() {
-	echo "Currently supported architectures:
+	echo "Currently supported architectures and operating systems:
 	- ppc64le
+		- ubuntu18
```
```suggestion
		- Ubuntu18
```

Haroon-Khel replied:
Ive kept the os lower case here as this is the way we want users to input the parameter using the -o flag

Willsparker replied:
The case statement you put in for the OSs covers the capital first letter too. Given the distributions are names, it looks weird to me to not have them capitalised

Haroon-Khel replied:
That sounds reasonable

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -115,12 +133,18 @@ defaultVars() {
 	- ppc64le
+		- ubuntu18
 	- s390x
+		- ubuntu18
```
```suggestion
		- Ubuntu18
```

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -115,12 +133,18 @@ defaultVars() {
 	- s390x
+		- ubuntu18
 	- arm32
+		- debian8
```
```suggestion
		- Debian8
```

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -115,12 +133,18 @@ defaultVars() {
+		- debian8
 	- aarch64
-	- riscv"
+		- debian10
```
```suggestion
		- Debian10
```

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -115,12 +133,18 @@ defaultVars() {
 	- aarch64
-	- riscv"
+		- debian10
+		- ubuntu18
```
```suggestion
		- Ubuntu18
```

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -115,12 +133,18 @@ defaultVars() {
+		- debian10
+		- ubuntu18
+	- riscv
+		- debian11"
```
```suggestion
		- Debian11"
```

Willsparker commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -81,19 +85,33 @@ usage() {
+		"ubuntu18" | "u18" | "Ubuntu18" )
+			echo "UBUNTU18 selected for $ARCHITECTURE"; OS=UBUNTU18;;
+		"debian11" | "deb11" | "Debian11" )
+			echo "DEBIAN11 selected for $ARCHITECTURE"; OS=DEBIAN11;;
```
```suggestion
			echo "Debian11 selected for \$ARCHITECTURE"; OS=DEBIAN11;;
```

Willsparker reviewed (changes requested):
Sorry to be so picky :joy:

sxa reviewed (commented):
When this gets merged let's make sure the final commit has a suitable prefix on it i.e. `pbTests:`

Haroon-Khel commented:
New qpc job running at https://ci.adoptopenjdk.net/job/QEMUPlaybookCheck/99/
Looks promising. All disk images boot and have the playbook run on them, except for ubuntu18.aarch64. This image worked fine when ran on `infra-softlayer-ubuntu1804-x64-1` on its own, but has trouble when run as part of the jenkins job.

Haroon-Khel commented:
@Willsparker fyi the riscv job has failed twice at 
```
12:48:39 TASK [Common : Add the openjdk repository to apt for openjdk7] *****************
12:48:51 fatal: [localhost]: FAILED! => {"changed": false, "cmd": "apt-key export DA1A4A13543B466853BAF164EB9B1D8886F44E2A", "msg": "E: gnupg, gnupg2 and gnupg1 do not seem to be installed, but one of them is required for this operation", "rc": 255, "stderr": "E: gnupg, gnupg2 and gnupg1 do not seem to be installed, but one of them is required for this operation\n", "stderr_lines": ["E: gnupg, gnupg2 and gnupg1 do not seem to be installed, but one of them is required for this operation"], "stdout": "", "stdout_lines": []}
```

Haroon-Khel commented:
ubuntu18.aarch64 disk wasnt booting on the jenkins job due to the jenkins user not permitted to access the image; fixed now. 
Also added a task to `Common/tasks/Ubuntu.yml` which installs `gnupg2` on aarch64; the playbook was getting caught up on `TASK [Common : Add the openjdk repository to apt]`, which failed due to `gnupg2` not being on the machine. This same fix may likely fix the riscv failure mentioned above

Haroon-Khel commented:
Latest qpc job running at https://ci.adoptopenjdk.net/job/QEMUPlaybookCheck/100/

Willsparker commented:
@Haroon-Khel The RISC-V playbook is yet to be merged (the fixes are at #1483).

Willsparker reviewed (approved):
LGTM

Willsparker commented:
@Haroon-Khel I checked to see why the aarch64 U18 platform kept failing in `QPC`, and I'm getting this error message when running the `qemu-system-aarch64` command manually: 
```
WARNING: Image format was not specified for '/home/vagrant4/U18_aaarch64_test/openjdk-infrastructure/ansible/pbTestScripts/qemu_pbCheck/UBUNTU18.AARCH64.dsk' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.
qemu-system-aarch64: Initialization of device cfi.pflash01 failed: Could not reopen file: Permission denied
```
When you specify the format to be raw, the first error message goes away, but there is still the permission denied error.

Haroon-Khel commented:
See that's odd, because I managed to get it working on that machine with the jenkins user by changing the owner of the `flash1.img` and `QEMU_EFI-flash.img` files in `/qemu_base_images/arm64_tools/` to the jenkins user

Haroon-Khel commented:
Now im getting a different permission error when booting the disk up as the jenkins user
```
qemu-system-aarch64: -device virtio-blk,drive=drive0,bootindex=0: Could not reopen file: Permission denied
```

Haroon-Khel commented:
Ok, ive got the qpc script to work as the jenkins user. I think this is a case of the machine needs to be disconnected and reconnected from jenkins in order to update its permissions. @sxa what do you think?

Haroon-Khel commented:
Disconnecting/reconnecting the machine from jenkins did not work. The job failed yet again at https://ci.adoptopenjdk.net/job/QEMUPlaybookCheck/ARCHITECTURE=aarch64,OS=ubuntu18,label=vagrant/109/console

Haroon-Khel commented:
@Willsparker despite the ubuntu18.arm64 job not working, can this still be merged? The changes to the qpc script are still vaild, since the other jobs run without error. Im eager to get this in because the qpc jenkins job is configured to use my fork and branch, which is holding up testing other changes to the qpc script, such as https://github.com/AdoptOpenJDK/openjdk-infrastructure/pull/1566

Haroon-Khel commented:
Changes the qpc job config back to original in order to test the pr https://github.com/AdoptOpenJDK/openjdk-infrastructure/pull/1566

Haroon-Khel commented:
It works! https://ci.adoptopenjdk.net/job/QEMUPlaybookCheck/ARCHITECTURE=aarch64,OS=ubuntu18,label=vagrant/120/console

Confusingly, though the jenkins user had read and write access and ownership of the files in `/qemu_base_images/arm64_tools/`, changing the permissions to `-rw-rw-rw-` seemed to do the trick.

@sxa can this finally be merged?

sxa commented:
Hmmm I'm not comfortable with anything that seems to require world-writable files around

Haroon-Khel commented:
Ive made the permissions of the contents of `/qemu_base_images/arm64_tools/` to `-rw-rw-r--` and the job fails as before, https://ci.adoptopenjdk.net/job/QEMUPlaybookCheck/122/ARCHITECTURE=aarch64,OS=ubuntu18,label=vagrant/console. 

In the logs, it seems that the jenkins node runs the job in the `/home/vagrantX` directory. and not in `/home/jenkins` as I thought. This may be why we are getting a permissions error.

How would you feel if I added the `vagrantX` users into the jenkins user group, on that machine? I think this may solve the problem

Haroon-Khel commented:
Ive moved the disk images from `/qemu_base_images` to `/home/jenkins/qemu_base_images`. Instead of copying over the compressed disk image, the script will copy the contents of a folder nameed OS.ARCHITECTURE, which will include the compressed disk image and any other files used in the booting of the vm, into the working directory, and then boot the image.
As of https://ci.adoptopenjdk.net/job/QEMUPlaybookCheck/136/console, this seems to work!

Willsparker reviewed (approved):
I like the compromise. All the platforms on QPC started the playbook fine. Could you document somewhere what the permission issues you were having were and what solved it? Probably in the _Setup-QEMU-Images.md_ doc. :-)

sxa commented:
Given that it's now doing the sensible thing and copying everything from the read-only repository and not writing directly into `/qemu_base_images` I'm not sure it's too necessary to add that into the documentation (which is unlike me to say something like that !)

sxa commented on ansible/pbTestScripts/qemuPlaybookCheck.sh:
```diff
@@ -137,9 +166,11 @@ setupWorkspace() {
 		rm -rf "$workFolder"/openjdk-infrastructure "$workFolder"/openjdk-build
 	fi
-	if [[ ! -f "${workFolder}/${ARCHITECTURE}.dsk" ]]; then 
+	if [[ ! -f "$workFolder/$OS.$ARCHITECTURE.dsk" ]]; then 
```
Why remove the braces around these variables?

Haroon-Khel replied:
I must have done it to make is consistent with the changes I made elsewhere in the script. Ive just tested it and either way works.
